mathics/core/list.py

```diff
@@ -58,6 +83,31 @@ def is_literal(self) -> bool:
         """
         return self._is_literal
 
+    def rewrite_apply_eval_step(self, evaluation) -> Tuple[Expression, bool]:
+        """
+        Perform a single rewrite/apply/eval step of the bigger
+        Expression.evaluate() process.
+
+        We return the ListExpression as well as a Boolean False which indicates
+        to the caller `evaluate()` that it should consider this a "fixed-point"
+        evaluation and not have to iterate calling this routine using the
+        returned results.
+
+        Note that this is a recursive process: we may call something
+        that may call our parent: evaluate() which calls us again.
+
+        In general, this step is time consuming, complicated, and involved.
+        However for a ListExpression, things are much much simpler and faster because
+        we don't need the rewrite/apply phases of evaluation.
+        """
+
+        if self.elements_properties is None:
+            self._build_elements_properties()
+
+        if not self.elements_properties.elements_fully_evaluated:
+            self.evaluate_elements(evaluation)
```

Comment:
@rocky, one possible problem with this line is that represents a change of behaviour regarding the same method in the superclass: in ``Expression`` this function does not modifies ``self``, but creates a new element and returns it.

A second problem is that this neglects upvalues. For example, consider the following code:
```
In[1]:= a={F[1],2};
In[2]:= {F[x_],t___}^:=Q[t];
In[3]:=a
```
the output must be `Q[2]`. However, with this implementation, the output would be `{F[x],2}`

Finally, the other problem is how ``Unevaluate`` elements are handled. This feature was broken for a while, but we should try to fix it eventually.

Reply:
> @rocky, one possible problem with this line is that represents a change of behaviour regarding the same method in the superclass: in `Expression` this function does not modifies `self`, but creates a new element and returns it.
> 

Ok. Good observation and one that can be easily fixed. What is a specific scenario where this important? 

> A second problem is that this neglects upvalues. For example, consider the following code:
> 
> ```
> In[1]:= a={F[1],2};
> In[2]:= {F[x_],t___}^:=Q[t];
> In[3]:=a
> ```
> 
> the output must be `Q[2]`. However, with this implementation, the output would be `{F[x],2}`
> 

Ok. I will add a test case for this and mark it "skipped" (not expected failure), and when we address this we can remove the skipped pytest tag.

> Finally, the other problem is how `Unevaluate` elements are handled. This feature was broken for a while, but we should try to fix it eventually. Here is completely lost.

Where specifically is this handled in `Expression.rewrite_apply_step` ?

Reply:
> 
> > Finally, the other problem is how `Unevaluate` elements are handled. This feature was broken for a while, but we should try to fix it eventually. Here is completely lost.
> 
> Where specifically is this handled in `Expression.rewrite_apply_step` ?

In `Expression.rewrite_apply_eval_step`, here: https://github.com/Mathics3/mathics-core/blob/5b2fb73854d76d6536182faa7e83d28d4b9c6a51/mathics/core/expression.py#L1223

In any case, this is broken by now. When I have the chance I am going to try to fix it.

Reply:
What are test-case examples that show the brokenness both for a general Expression and a ListExpression. 

But at any rate, I am happy to hold off looking at this if you're going to look at. Basically we want to get the Expression part closed out so we can release (hopefully by July which is when the next round of NumFocus adjunct applications starts). And then after the release you'll have free ground in working on Boxing and Formatting.  

Going over the rewrite and apply phases, I guess will come after that.